### PR TITLE
add --rebuild flag to client and server

### PR DESCRIPTION
**Pull request by dtheodor**
### Rebuilding a single image, from the client:

`mistry build --project foo --rebuild`

Add a new boolean `JobRequest.Rebuild`, which passes the options `--pull` and `--no-cache` to the image build.

Also add `--force-rm` by default true to image build, to avoid loose containers on build failures

### Rebuilding on the server

`mistryd --config <file> rebuild --project foo --project bar --fail-fast`

* can rebuild specific projects, or all of them if no projects are passed
* fail-fast whether to exit on first error, or keep running the rest of the images

### CODE REVIEW

File: cmd/mistry/main.go [hunk 1]
```diff
@@ -121,6 +122,11 @@ EXAMPLES:
 					Usage:       "output the build result in JSON format to STDOUT (implies verbose: false)",
 					Destination: &jsonResult,
 				},
+				cli.BoolFlag{
+					Name:        "rebuild",
+					Usage:       "whether to completely rebuild the docker image",
```
Comment (agis) [inline cmd/mistry/main.go]: s/whether to//

File: cmd/mistryd/end_to_end_test.go [hunk 2]
```diff
@@ -24,6 +24,13 @@ func TestSimpleBuild(t *testing.T) {
 	}
 }
 
+func TestSimpleRebuild(t *testing.T) {
+	cmdout, cmderr, err := cliBuildJob("--project", "simple", "--rebuild")
+	if err != nil {
+		t.Fatalf("mistry-cli stdout: %s, stderr: %s, err: %#v", cmdout, cmderr, err)
+	}
+}
```
Comment (agis) [inline cmd/mistryd/end_to_end_test.go]: You could probably inspect the image and test if it _actually_ was rebuilt.

File: cmd/mistryd/job.go [hunk 3]
```diff
@@ -132,10 +132,17 @@ func NewJob(project string, params types.Params, group string, cfg *Config) (*Jo
 
 // BuildImage builds the Docker image denoted by j.Image. If there is an
 // error, it will be of type types.ErrImageBuild.
-func (j *Job) BuildImage(ctx context.Context, uid string, c *docker.Client, out io.Writer) error {
+func (j *Job) BuildImage(ctx context.Context, uid string, c *docker.Client, out io.Writer, pullParent bool, noCache bool) error {
```
Comment (agis) [inline cmd/mistryd/job.go]: You could leave out the first "bool".

File: cmd/mistryd/job.go [hunk 4]
```diff
@@ -132,10 +132,17 @@ func NewJob(project string, params types.Params, group string, cfg *Config) (*Jo
...
 	buildArgs := make(map[string]*string)
 	buildArgs["uid"] = &uid
-	buildOpts := dockertypes.ImageBuildOptions{Tags: []string{j.Image}, BuildArgs: buildArgs, NetworkMode: "host"}
+	buildOpts := dockertypes.ImageBuildOptions{
+		Tags:        []string{j.Image},
+		BuildArgs:   buildArgs,
+		NetworkMode: "host",
+		PullParent:  pullParent,
+		NoCache:     noCache,
+		ForceRemove: true,
```
Comment (agis) [inline cmd/mistryd/job.go]: What does this actually do? Doesn't this affect the image caching layers?

File: pkg/types/job_request.go [hunk 5]
```diff
@@ -1,7 +1,9 @@
 package types
 
+// JobRequest contains the data the job was requested with
 type JobRequest struct {
 	Project string
 	Params  Params
 	Group   string
+	Rebuild bool
 }
```
Comment (agis) [inline pkg/types/job_request.go]: [out of scope] doesn't this belong as a field inside `Job{}`?

Comment (agis) [reply pkg/types/job_request.go]: Nvm this.

File: cmd/mistryd/main.go [hunk 6]
```diff
@@ -114,22 +139,28 @@ func main() {
...
+	err := PruneZombieBuilds(cfg)
 	if err != nil {
 		return err
 	}
 
-	err = utils.PathIsDir(cfg.BuildPath)
+	return nil
+}
+
+func parseAndValidateConfig(c *cli.Context) (*Config, error) {
```
Comment (agis) [inline cmd/mistryd/main.go]: could just be "parseConfig"

File: cmd/mistryd/main.go [hunk 7]
```diff
@@ -84,27 +83,53 @@ func main() {
...
+			Name:  "rebuild",
+			Usage: "Rebuild docker images for all projects.",
+			Flags: []cli.Flag{
+				cli.BoolFlag{
+					Name:  "fail-fast",
+					Usage: "exit immediately on first image error",
+				},
+				cli.StringSliceFlag{
+					Name:  "project, p",
+					Usage: "the project to build. multiple projects can be specified. if not passed, all projects are built",
```
Comment (agis) [inline cmd/mistryd/main.go]: [nit] capitalize words after periods

File: cmd/mistryd/main.go [hunk 8]
```diff
@@ -84,27 +83,53 @@ func main() {
...
 		return StartServer(cfg)
 	}
+	app.Commands = []cli.Command{
+		{
+			Name:  "rebuild",
+			Usage: "Rebuild docker images for all projects.",
+			Flags: []cli.Flag{
+				cli.BoolFlag{
+					Name:  "fail-fast",
+					Usage: "exit immediately on first image error",
```
Comment (agis) [inline cmd/mistryd/main.go]: s/image//

**Comment by dtheodor** [comment]
going to add scheduling the rebuild in this PR as well

File: cmd/mistryd/server.go [hunk 9]
```diff
@@ -402,25 +406,121 @@ func (s *Server) ListenAndServe() error {
...
+	}
+	return nil
+}
+
+func getProjects(path string) ([]string, error) {
+	folders, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	projects := []string{}
```
Comment (dtheodor) [inline cmd/mistryd/server.go]: should use `make([]string, len(folders))` to build this slice

Comment (agis) [reply cmd/mistryd/server.go]: What was the problem with just a simple slice without len? Why is the the trick mentioned in the link needed in our case?

Comment (dtheodor) [reply cmd/mistryd/server.go]: just a micro optimization

Comment (dtheodor) [reply cmd/mistryd/server.go]: not doing it

File: cmd/mistryd/main.go (hunk 6 above)
Comment (dtheodor) [reply cmd/mistryd/main.go]: the function `ParseConfig` already exists and parses the JSON, this one calls it and validates results

maybe it should just be 1 function

Comment (dtheodor) [reply cmd/mistryd/main.go]: n/m I'm renaming to `parseConfigFromCli`. it knows how to turn CLI flags into a proper config. the `ParseConfig` function remains, but knows nothing about the CLI flags

**Pull request merged**